File: mobile_apps/Delivera Swift iOS/DeliveraSwift/Features/Home/HomeViewModel.swift
import Foundation

@MainActor
final class HomeViewModel: ObservableObject {
    @Published private(set) var restaurants: [Restaurant] = []
    @Published private(set) var sponsors: [Sponsor] = []
    @Published private(set) var trackingAds: [TrackingAd] = []
    @Published private(set) var sections: [HomeCategorySection] = []
    @Published private(set) var cities: [City] = []
    @Published private(set) var isLoading = true
    @Published var errorMessage: String?

    private let api = DeliveraAPI()

    var cuisines: [String] {
        let values = restaurants.flatMap { Self.cuisineTokens(from: $0.cuisine) }
        return ["Alla"] + Array(Set(values)).sorted().prefix(8)
    }

    func load() async {
        isLoading = true
        errorMessage = nil

        async let restaurantRequest = result { try await api.restaurants() }
        async let sponsorRequest = result { try await api.sponsors() }
        async let adRequest = result { try await api.trackingAds() }
        async let sectionRequest = result { try await api.homeSections() }
        async let cityRequest = result { try await api.cities() }

        let restaurantResult = await restaurantRequest
        let sponsorResult = await sponsorRequest
        let adResult = await adRequest
        let sectionResult = await sectionRequest
        let cityResult = await cityRequest

        if case .success(let nextRestaurants) = restaurantResult {
            restaurants = nextRestaurants.sortedForHome()
        }
        if case .success(let nextSponsors) = sponsorResult {
            sponsors = nextSponsors.filter { $0.isActive != false }
        }
        if case .success(let nextAds) = adResult {
            trackingAds = nextAds
                .filter { $0.isActive != false }
                .sorted { ($0.sortOrder ?? 0) < ($1.sortOrder ?? 0) }
        }
        if case .success(let nextSections) = sectionResult {
            sections = nextSections.filter { $0.isActive }
        }
        if case .success(let nextCities) = cityResult {
            cities = nextCities.filter { $0.isActive }
        }

        if restaurants.isEmpty, case .failure = restaurantResult {
            errorMessage = "Kunde inte hämta data. Kontrollera att API:t kör på samma Wi-Fi."
        } else {
            errorMessage = nil
        }

        isLoading = false
    }

    func filteredRestaurants(cuisine: String, searchQuery: String, cityName: String?) -> [Restaurant] {
        filter(restaurants, cuisine: cuisine, searchQuery: searchQuery, cityName: cityName)
    }

    func restaurants(for section: HomeCategorySection, cityName: String?) -> [Restaurant] {
        var result = restaurants

        if !section.manualRestaurantIds.isEmpty {
            let ids = Set(section.manualRestaurantIds)
            let manual = restaurants.filter { ids.contains($0.id) }
            result = section.filterMode == "MANUAL" ? manual : (manual + result)
        }

        if let searchTerm = section.filters.searchTerm?.lowercased(), !searchTerm.isEmpty {
            result = result.filter {
                $0.name.lowercased().contains(searchTerm) ||
                ($0.cuisine?.lowercased().contains(searchTerm) ?? false) ||
                ($0.tags ?? []).contains { $0.lowercased().contains(searchTerm) }
            }
        }

        if !section.filters.cuisines.isEmpty {
            let cuisines = Set(section.filters.cuisines.map { $0.lowercased() })
            result = result.filter { cuisines.contains(($0.cuisine ?? "").lowercased()) }
        }

        if !section.filters.featuredClasses.isEmpty {
            let classes = Set(section.filters.featuredClasses)
            result = result.filter { classes.contains($0.featuredClass ?? 99) }
        }

        if let maxEta = section.filters.maxEtaMinutes {
            result = result.filter { ($0.etaMinutes ?? 30) <= maxEta }
        }

        if section.filters.freeDeliveryOnly {
            result = result.filter { ($0.deliveryFee ?? 0) <= 0 }
        }

        if section.filters.openNowOnly {
            result = result.filter { $0.isOpen != false }
        }

        result = filter(result, cuisine: "Alla", searchQuery: "", cityName: cityName)

        return Array(result.uniqued().sortedForHome().prefix(section.maxRestaurants))
    }

    private func filter(_ source: [Restaurant], cuisine: String, searchQuery: String, cityName: String?) -> [Restaurant] {
        var result = source

        if let cityName, !cityName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            result = result.filter { restaurant in
                guard let city = restaurant.city else { return true }
                return city.localizedCaseInsensitiveContains(cityName) || cityName.localizedCaseInsensitiveContains(city)
            }
        }

        if cuisine != "Alla" {
            result = result.filter { restaurant in
                let cuisines = Self.cuisineTokens(from: restaurant.cuisine)
                return cuisines.contains { $0.localizedCaseInsensitiveCompare(cuisine) == .orderedSame } ||
                    (restaurant.tags ?? []).contains { $0.localizedCaseInsensitiveContains(cuisine) }
            }
        }

        let query = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        if !query.isEmpty {
            result = result.filter { restaurant in
                restaurant.name.localizedCaseInsensitiveContains(query) ||
                (restaurant.cuisine?.localizedCaseInsensitiveContains(query) ?? false) ||
                (restaurant.tags ?? []).contains { $0.localizedCaseInsensitiveContains(query) }
            }
        }

        return result.sortedForHome()
    }

    private static func cuisineTokens(from value: String?) -> [String] {
        guard let value else { return [] }
        let normalized = value
            .replacingOccurrences(of: " och ", with: ",", options: [.caseInsensitive])
            .replacingOccurrences(of: "/", with: ",")
            .replacingOccurrences(of: "&", with: ",")
        return normalized
            .split(separator: ",")
            .map { token in
                token.trimmingCharacters(in: .whitespacesAndNewlines).capitalized
            }
            .filter { !$0.isEmpty }
    }

    private func result<T>(_ operation: () async throws -> T) async -> Result<T, Error> {
        do {
            return .success(try await operation())
        } catch {
            return .failure(error)
        }
    }
}

File: mobile_apps/Delivera Android/app/src/main/java/se/delivera/android/ui/home/HomeViewModel.kt
package se.delivera.android.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import se.delivera.android.data.City
import se.delivera.android.data.DeliveraApi
import se.delivera.android.data.HomeAppDeal
import se.delivera.android.data.HomeCategorySection
import se.delivera.android.data.HomePulseModule
import se.delivera.android.data.Restaurant
import se.delivera.android.data.Sponsor
import se.delivera.android.data.TrackingAd
import se.delivera.android.data.cuisineTokens
import se.delivera.android.data.sortedForHome
import se.delivera.android.data.uniqued

data class HomeState(
    val restaurants: List<Restaurant> = emptyList(),
    val sponsors: List<Sponsor> = emptyList(),
    val appDeals: List<HomeAppDeal> = emptyList(),
    val pulseModules: List<HomePulseModule> = emptyList(),
    val greeting: String? = null,
    val trackingAds: List<TrackingAd> = emptyList(),
    val sections: List<HomeCategorySection> = emptyList(),
    val cities: List<City> = emptyList(),
    val isLoading: Boolean = true,
    val errorMessage: String? = null
)

/** Port of HomeViewModel.swift, including filter/section logic. */
class HomeViewModel : ViewModel() {
    private val api = DeliveraApi()

    private val _state = MutableStateFlow(HomeState())
    val state = _state.asStateFlow()

    val cuisines: List<String>
        get() {
            val values = _state.value.restaurants.flatMap { cuisineTokens(it.cuisine) }
            return listOf("Alla") + values.toSortedSet().take(8)
        }

    fun load(authToken: String = "") {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, errorMessage = null)
            val trimmed = authToken.trim()
            val isLoggedIn = trimmed.isNotEmpty()

            val restaurantReq = async { runCatching { api.restaurants() } }
            val sponsorReq = async { runCatching { api.sponsors() } }
            val dealReq = async { runCatching { api.homeAppDeals(isLoggedIn, trimmed.ifEmpty { null }) } }
            val pulseReq = async { runCatching { api.homePulse(trimmed.ifEmpty { null }) } }
            val adReq = async { runCatching { api.trackingAds() } }
            val sectionReq = async { runCatching { api.homeSections() } }
            val cityReq = async { runCatching { api.cities() } }

            val restaurantRes = restaurantReq.await()
            val sponsorRes = sponsorReq.await()
            val dealRes = dealReq.await()
            val pulseRes = pulseReq.await()
            val adRes = adReq.await()
            val sectionRes = sectionReq.await()
            val cityRes = cityReq.await()

            var s = _state.value
            restaurantRes.getOrNull()?.let { s = s.copy(restaurants = it.sortedForHome()) }
            sponsorRes.getOrNull()?.let { list -> s = s.copy(sponsors = list.filter { it.isActive != false }) }
            dealRes.getOrNull()?.let { s = s.copy(appDeals = it.deals) }
            pulseRes.getOrNull()?.let { s = s.copy(pulseModules = it.modules, greeting = it.greeting) }
            adRes.getOrNull()?.let { list ->
                s = s.copy(trackingAds = list.filter { it.isActive != false }.sortedBy { it.sortOrder ?: 0 })
            }
            sectionRes.getOrNull()?.let { list -> s = s.copy(sections = list.filter { it.isActive }) }
            cityRes.getOrNull()?.let { list -> s = s.copy(cities = list.filter { it.isActive }) }

            s = if (s.restaurants.isEmpty() && restaurantRes.isFailure) {
                s.copy(errorMessage = "Kunde inte hämta data. Kontrollera din anslutning.")
            } else {
                s.copy(errorMessage = null)
            }
            _state.value = s.copy(isLoading = false)
        }
    }

    fun replaceDeal(deal: HomeAppDeal) {
        val deals = _state.value.appDeals.toMutableList()
        val index = deals.indexOfFirst { it.id == deal.id }
        if (index == -1) return
        deals[index] = deal
        _state.value = _state.value.copy(appDeals = deals)
    }

    fun removeDeal(id: String) {
        _state.value = _state.value.copy(appDeals = _state.value.appDeals.filterNot { it.id == id })
    }

    fun filteredRestaurants(cuisine: String, searchQuery: String, cityName: String?): List<Restaurant> =
        filter(_state.value.restaurants, cuisine, searchQuery, cityName)

    fun restaurantsFor(section: HomeCategorySection, cityName: String?): List<Restaurant> {
        var result = _state.value.restaurants

        if (section.manualRestaurantIds.isNotEmpty()) {
            val ids = section.manualRestaurantIds.toSet()
            val manual = result.filter { ids.contains(it.id) }
            result = if (section.filterMode == "MANUAL") manual else manual + result
        }

        section.filters.searchTerm?.lowercase()?.takeIf { it.isNotEmpty() }?.let { term ->
            result = result.filter {
                it.name.lowercase().contains(term) ||
                    (it.cuisine?.lowercase()?.contains(term) ?: false) ||
                    (it.tags ?: emptyList()).any { t -> t.lowercase().contains(term) }
            }
        }

        if (section.filters.cuisines.isNotEmpty()) {
            val cuisines = section.filters.cuisines.map { it.lowercase() }.toSet()
            result = result.filter { cuisines.contains((it.cuisine ?: "").lowercase()) }
        }

        if (section.filters.featuredClasses.isNotEmpty()) {
            val classes = section.filters.featuredClasses.toSet()
            result = result.filter { classes.contains(it.featuredClass ?: 99) }
        }

        section.filters.maxEtaMinutes?.let { maxEta ->
            result = result.filter { (it.etaMinutes ?: 30) <= maxEta }
        }

        if (section.filters.freeDeliveryOnly) result = result.filter { (it.deliveryFee ?: 0.0) <= 0 }
        if (section.filters.openNowOnly) result = result.filter { it.isOpen != false }

        result = filter(result, "Alla", "", cityName)
        return result.uniqued().sortedForHome().take(section.maxRestaurants)
    }

    private fun filter(source: List<Restaurant>, cuisine: String, searchQuery: String, cityName: String?): List<Restaurant> {
        var result = source
        if (!cityName.isNullOrBlank()) {
            result = result.filter { r ->
                val city = r.city ?: return@filter true
                city.contains(cityName, ignoreCase = true) || cityName.contains(city, ignoreCase = true)
            }
        }
        if (cuisine != "Alla") {
            result = result.filter { r ->
                val tokens = cuisineTokens(r.cuisine)
                tokens.any { it.equals(cuisine, ignoreCase = true) } ||
                    (r.tags ?: emptyList()).any { it.contains(cuisine, ignoreCase = true) }
            }
        }
        val query = searchQuery.trim()
        if (query.isNotEmpty()) {
            result = result.filter { r ->
                r.name.contains(query, ignoreCase = true) ||
                    (r.cuisine?.contains(query, ignoreCase = true) ?: false) ||
                    (r.tags ?: emptyList()).any { it.contains(query, ignoreCase = true) }
            }
        }
        return result.sortedForHome()
    }
}
